src/Interpreters/Context.cpp

```diff
@@ -1201,44 +1214,47 @@ String Context::getFilesystemCacheUser() const
     return shared->filesystem_cache_user;
 }
 
-Strings Context::getWarnings() const
+std::unordered_map<Context::WarningType, Context::Warning> Context::getWarnings() const
 {
-    Strings common_warnings;
+    std::unordered_map<Context::WarningType, Context::Warning> common_warnings;
     {
         SharedLockGuard lock(shared->mutex);
         common_warnings = shared->warnings;
+        constexpr auto message_format_string = "The number of attached {} is more than {}.";
         if (CurrentMetrics::get(CurrentMetrics::AttachedTable) > static_cast<Int64>(shared->max_table_num_to_warn))
-            common_warnings.emplace_back(fmt::format("The number of attached tables is more than {}.", shared->max_table_num_to_warn));
+            common_warnings[Context::WarningType::MAX_ATTACHED_TABLES] = Context::Warning{
+                .message = fmt::format(message_format_string, "tables", shared->max_table_num_to_warn),
+                .message_format_string = message_format_string};
         if (CurrentMetrics::get(CurrentMetrics::AttachedView) > static_cast<Int64>(shared->max_view_num_to_warn))
-            common_warnings.emplace_back(fmt::format("The number of attached views is more than {}.", shared->max_view_num_to_warn));
+            common_warnings[Context::WarningType::MAX_ATTACHED_VIEWS] = Context::Warning{
+                .message = fmt::format(message_format_string, "views", shared->max_view_num_to_warn),
+                .message_format_string = message_format_string};
         if (CurrentMetrics::get(CurrentMetrics::AttachedDictionary) > static_cast<Int64>(shared->max_dictionary_num_to_warn))
-            common_warnings.emplace_back(fmt::format("The number of attached dictionaries is more than {}.", shared->max_dictionary_num_to_warn));
+            common_warnings[Context::WarningType::MAX_ATTACHED_DICTIONARIES] = Context::Warning{
+                .message = fmt::format(message_format_string, "dictionaries", shared->max_dictionary_num_to_warn),
+                .message_format_string = message_format_string};
         if (CurrentMetrics::get(CurrentMetrics::AttachedDatabase) > static_cast<Int64>(shared->max_database_num_to_warn))
-            common_warnings.emplace_back(fmt::format("The number of attached databases is more than {}.", shared->max_database_num_to_warn));
+            common_warnings[Context::WarningType::MAX_ATTACHED_DATABASES] = Context::Warning{
+                .message = fmt::format(message_format_string, "databases", shared->max_database_num_to_warn),
+                .message_format_string = message_format_string};
```

Comment:
I don't think message_format_string needs "tables", "views" as parameter - doesn't make much sense in table - better to have individual string for each.